### ISSUE TITLE
Cookbook incompatible with Chef 13 (deprecation warning)

### ISSUE THREAD

**Issue by wolf31o2** [treasure-data/chef-td-agent#112]
When using this cookbook under Chef 12 w/ deprecation warnings as errors enabled, we get this:

```
Compiling Cookbooks...
[2018-02-05T16:15:08+00:00] ERROR: Property `params` of resource `td_agent_filter` overwrites an existing method. Please use a different property name. This will raise an exception in Chef 13. (CHEF-11)/opt/kitchen/cache/cookbooks/td-agent/resources/filter.rb:33:in `class_from_file'.
Please see https://docs.chef.io/deprecations_property_name_collision.html for further details and information on how to correct this problem. at /opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.3/lib/chef/formatters/doc.rb:419:in `deprecation'

================================================================================
Recipe Compile Error in /opt/kitchen/cache/cookbooks/td-agent/resources/filter.rb
================================================================================

Chef::Exceptions::DeprecatedFeatureError
----------------------------------------
"Property `params` of resource `td_agent_filter` overwrites an existing method. Please use a different property name. This will raise an exception in Chef 13. (CHEF-11)/opt/kitchen/cache/cookbooks/td-agent/resources/filter.rb:33:in `class_from_file'.\nPlease see https://docs.chef.io/deprecations_property_name_collision.html for further details and information on how to correct this problem. at /opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.3/lib/chef/formatters/doc.rb:419:in `deprecation'"

Cookbook Trace:
---------------
  /opt/kitchen/cache/cookbooks/td-agent/resources/filter.rb:33:in `class_from_file'

Relevant File Content:
----------------------
/opt/kitchen/cache/cookbooks/td-agent/resources/filter.rb:

 26:  attribute :type, :kind_of => String, :required => true
 27:  attribute :tag, :kind_of => String, :required => true
 28:  attribute :parameters, :default => {}
 29:
 30:  # Workaround for backward compatibility for Chef pre-13 (#99)
 31:  chef_major_version = ::Chef::VERSION.split(".").first.to_i
 32:  if chef_major_version < 13
 33>>   attribute :params, :default => {}
 34:  end
 35:

System Info:
------------
chef_version=12.21.3
platform=centos
platform_version=7.4.1708
ruby=ruby 2.3.1p112 (2016-04-26 revision 54768) [x86_64-linux]
program_name=chef-client worker: ppid=201;start=16:14:47;
executable=/opt/chef/embedded/bin/chef-client


Running handlers:
[2018-02-05T16:15:08+00:00] ERROR: Running exception handlers
Running handlers complete
[2018-02-05T16:15:08+00:00] ERROR: Exception handlers complete
Chef Client failed. 0 resources updated in 21 seconds
[2018-02-05T16:15:08+00:00] FATAL: Stacktrace dumped to /opt/kitchen/cache/chef-stacktrace.out
[2018-02-05T16:15:08+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2018-02-05T16:15:08+00:00] ERROR: "Property `params` of resource `td_agent_filter` overwrites an existing method. Please use a different property name. This will raise an exception in Chef 13. (CHEF-11)/opt/kitchen/cache/cookbooks/td-agent/resources/filter.rb:33:in `class_from_file'.\nPlease see https://docs.chef.io/deprecations_property_name_collision.html for further details and information on how to correct this problem. at /opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.3/lib/chef/formatters/doc.rb:419:in `deprecation'"
[2018-02-05T16:15:09+00:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
```

It looks like some work was done to work around this, but I am still seeing this.

**Comment by yyuu** [comment]
The workaround has been added in #93 and #99, and left `params` for backward compatibility with 12.x. However, apparently starting from 12.19, it started refusing the use of `params` name. The code around there won't work with these newer 12.x releases. I'm going to prepare some workaround for it later.

**Comment by yyuu** [comment]
I realized that we're running CI with using 12.21.3 + `params` keyword and it succeeded without raising `Chef::Exceptions::DeprecatedFeatureError`.

https://github.com/treasure-data/chef-td-agent/blob/54cdd6434af3282fa09ef498ec7a6a71d7947f19/.kitchen.yml#L45
https://travis-ci.org/treasure-data/chef-td-agent/jobs/336439478

Could you provide some details in your procedures to reproduce the error...?

**Comment by wolf31o2** [comment]
I'm using client.rb options to treat deprecations as errors. I'm guessing that Chef is detecting the deprecation even though you've guarded against it with a version check at runtime.

**Comment by yyuu** [comment]
The option you mean is `deprecations_as_errors` in test-kitchen (=~ `treat_deprecation_warnings_as_errors` in chef)...? It's clear we're not enabling the feature in CI.

https://docs.chef.io/chef_deprecations_client.html

**Comment by wolf31o2** [comment]
Correct. We use it in all of our CI and this cookbook (currently) requires me to disable it.

**Comment by yyuu** [comment]
I've made some fix as #113. Please verify if it works for you. Once you confirmed, I'm going to release it to Chef Supermarket.

**Comment by wolf31o2** [comment]
This has been resolved. Please, release a new version to Chef Supermarket. 😃

**Comment by yyuu** [comment]
Released 3.0.2

https://supermarket.chef.io/cookbooks/td-agent/versions/3.0.2
